### PR TITLE
Add possibility to set the alternate pin mode on Raspberry Pis

### PR DESCRIPTION
**Pull request by pgrawehr**
This is a split from PR #1128 

CC: @krwq, @joperezr

### CODE REVIEW

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs:360]: argument name should be `extendedPinMode` here and in any similar place like this

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPiExtendedPinMode.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPiExtendedPinMode.cs:1]: copyright header

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs:426]: since you have exactly 8 values it's faster to create an array with all entries:
`ExtendedPinMode[] s_pinModes = new[8] { ... };`

and then just use indexer here

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs:426]: you can put this array in the RaspberryPiExtendedPinMode and have a method in there which resolves int into the instance

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: I'd refrain from doing IEquatable for now unless we have a clear use case.

Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: I'd suggest abstract

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:22]: for unknown just create an internal class UnknownExtendedPinMode with single instance

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:28]: I'd suggest to do this argumentless and make implementation choose if they want to use string or something else to get a name (i.e. in RaspberryPiExtendedMode you don't really need any string)

File: src/System.Device.Gpio/System/Device/Gpio/GpioController.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/GpioController.cs:244]: GetExtendedPinMode (same comment for Set*)

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:22]: I need this to be public, since clients (in this case the board class) needs to compare against it.

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: I need to create the "Unknown" instance below, therefore this is not abstract. Since the ctor is protected, the effect is the same.

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:28]: I know that the string is not absolutely necessary, but it helps in keeping the things testable. Also, otherwise the instance wouldn't necessarily have any identity at all and comparing them was impossible.

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: The user needs to do comparisons (i.e. `if (mode == Modes.GPIO)...`). Without implementing IEquatable, things get needlessly complicated.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs:426]: Not sure this is a good idea. For one, this would cause a static initializer depenceny between this class and RaspberryPi3Driver, and then it may be confusing, because the order is not sequential.

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (Ellerbach) [inline src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:59]: super minor but you may want to use the condense notation => for anything that just do a return or call a single function. Apply to this and the next 3 ones

**Review by Ellerbach — APPROVED** [review]
I'm fine with this implementation. Would be good to add a readme somewhere to demonstrate to use and a sample code. That would make things easier. 
I'm thinking of checking if a pin is PWM, changing one it to PWM and creating a hardware PWM then changing things back.

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:22]: This still can be public, UnknownExtendedPinMode  would be internal and ExtendedPinMode decoupled from `string`

File: src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: There is no problem with having static readonly members on abstract class

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: This will still work considering you return the same reference

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: Sure, the problem is not having the static readonly member, but constructing its instance. The `UnknownExtendedPinMode` class would work around this, but I don't really see the benefit of an additional class just to make this abstract.

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/ExtendedPinMode.cs:15]: You are right, that would work with reference comparison as well. But considering that the class is inheritable, it should be made clear that instances of it are meant to be compared against each other. Maybe some derived implementations will not use statics, but dynamically create the instances (i.e. because the set of possible modes is only known at runtime).

**Comment by pgrawehr** [comment]
> 
> 
> I'm fine with this implementation. Would be good to add a readme somewhere to demonstrate to use and a sample code. That would make things easier.
> I'm thinking of checking if a pin is PWM, changing one it to PWM and creating a hardware PWM then changing things back.

While it is of course possible to use this directly, use of the higher-level board abstraction to access these functions is clearly preferred. There should be documentation about all this there already, iirc.

**Comment by krwq** [comment]
cc: @microhobby (ref: https://github.com/dotnet/iot/pull/1128#discussion_r550335051)

I'm curious on how else can we solve issues like https://github.com/dotnet/iot/issues/874

@microhobby Do you also suggest to not have memory based implementation of GpioDriver? I personally like being able to change mode of the pin in the app rather than having changing global config file and reboot.

**Comment by microhobby** [comment]
> cc: @microhobby (ref: [#1128 (comment)](https://github.com/dotnet/iot/pull/1128#discussion_r550335051))
> 
> I'm curious on how else can we solve issues like #874
>

I don't know how the tests are implemented, but I believe for the RPI we are using  `/dev/mem` and so they are changing the alternative mode to GPIO, overwriting the mode that is defined on device tree and configured by the Kernel on the boot, which should lead to PWMs not working. But I'm speculating I should take a look at the implementation to say for sure. This is something that for example `gpiolib` is concerned about, will we change the state of a pin? Ok, but as soon as the process that changed the state is done or the pin is put into disuse it goes back to its reset state.
 
> @microhobby Do you also suggest to not have memory based implementation of GpioDriver? I personally like being able to change mode of the pin in the app rather than having changing global config file and reboot.

I agree that it is easier to be able to change the pin alternative modes in user space, without the need for a reboot or possible driver reload. But this is not the recommended way. It can be useful for development but should not be encouraged for production. For production the use of devices trees or devices trees overlays are the recommended way. The Kernel knows the hardware by the description given by the device tree, it is the Kernel that knows and will best handle these resources. For this case `gpiolib` will list the pin with an alternate function set as different from GPIO as used by the kernel and will not attempt to change it.

I am also curious if for example we change the alternative function of an i2c pin without it having been previously defined in the device tree, if it will be listed in the i2c sysfs and if after changing the registers by `/dev/mem` the driver will continue to work or whether we will have to write our own i2c implementations. 🤷‍♂️ I've never tried anything like this.

**Comment by pgrawehr** [comment]
> I agree that it is easier to be able to change the pin alternative modes in user space, without the need for a reboot or possible driver reload. But this is not the recommended way. It can be useful for development but should not be encouraged for production. For production the use of devices trees or devices trees overlays are the recommended way. The Kernel knows the hardware by the description given by the device tree, it is the Kernel that knows and will best handle these resources. For this case `gpiolib` will list the pin with an alternate function set as different from GPIO as used by the kernel and will not attempt to change it.
> 
> I am also curious if for example we change the alternative function of an i2c pin without it having been previously defined in the device tree, if it will be listed in the i2c sysfs and if after changing the registers by `/dev/mem` the driver will continue to work or whether we will have to write our own i2c implementations. 🤷‍♂️ I've never tried anything like this.

I haven't tested it with I2C, but from my tests with PWM I don't expect anything "bad" to happen. I'm actually expecting that you could just enable all drivers you might possibly need in `boot/config.txt` and then switch at runtime to whichever you need. Of course, if you enable i.e. UART and I2C on pins 0&1, you'll need an external multiplexer to actually use them both.

Just tested this scenario. I added these two lines to `boot/config.txt`:
```
dtoverlay=i2c0 # uses pins 0 & 1, mode ALT0
dtoverlay=uart2 # also uses pins 0 & 1, mode ALT4
```

So now I could switch the pins between modes 0 and 4 to change the modes, the kernel was all fine with it it seems, and `i2cdetect -y 0` would work mostly fine. There was, however, some interference to be observed. When I changed the pins to ALT0 and ran that command, it would print all addresses as used despite there's nothing connected, really. That seems to be more of a hardware issue than a driver problem though and is out of scope here.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:29]: - make them public rather than protected

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs]: @krwq, @joperezr Updated PR according to our discussion. I did have to add a property to query the support, because it doesn't currently work on Windows, and I don't necessarily want to rely on exceptions in this case. (And Windows 10 IOT support on the Raspi is dying out it seems, anyway)

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:164]: This should be public as well, otherwise user can't check it

Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:164]: actually you akready have AlternatePinModeSettingSupported, this should be removed in that case

Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:164]: True, this was just forgotten.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:131]: please only leave this one

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (krwq) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:164]: I'm still seeing this

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:37]: Since you never return this or allow this to set, perhaps consider removing this value

**Comment by krwq** [comment]
@pgrawehr still seeing ExtendedPinModeSupported.

**Review by krwq — APPROVED** [review]
LGTM after 1-2 fixes:
- consider removing AltMode.Unknown since you never use it
- remove ExtendedPinModeSupported

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs
Comment (pgrawehr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs:37]: I found that it will be helpful for the client, therefore I added it.

**Comment by krwq** [comment]
@microhobby just FYI we have changed the implementation to limit this functionality to only the driver which already uses direct memory access, we do not plan to have a generic functionality like was initially proposed.

**Review by krwq — APPROVED** [review]

**Pull request merged**
